libraries/ms-common/src/stm32f0xx/adc.c

```diff
@@ -0,0 +1,183 @@
+#include <stdbool.h>
+
+#include "adc.h"
+#include "log.h"
+#include "stm32f0xx.h"
+
+#define TS_CAL1 0x1FFFF7b8
+#define TS_CAL2 0x1FFFF7c2
+#define VREFINT_CAL 0x1FFFF7ba
+
+typedef struct ADCInterrupt {
+  ADCCallback callback;
+  void *context;
+  uint16_t reading;
+} ADCInterrupt;
+
+typedef struct ADCStatus {
+  uint32_t sequence;
+  bool continuous;
+} ADCStatus;
+
+static ADCInterrupt s_adc_interrupts[NUM_ADC_CHANNEL];
+static ADCStatus s_adc_status;
+
+static uint16_t prv_get_temp(uint16_t reading) {
+  uint16_t ts_cal1 = *(uint16_t*)TS_CAL1;
+  uint16_t ts_cal2 = *(uint16_t*)TS_CAL2;
+
+  reading = ((110 - 30) * (reading - ts_cal1)) / (ts_cal2 - ts_cal1) + 30;
+
+  return reading + 273;
+}
+
+static uint16_t prv_get_vdda(uint16_t reading) {
+  uint16_t vrefint_cal = *(uint16_t*)VREFINT_CAL;
+  reading = (3300 * vrefint_cal) / reading;
+  return reading;
+}
+
+void adc_init(ADCMode adc_mode) {
+  ADC_DeInit(ADC1);
+
+  // Once the ADC has been reset, enable it with the given settings
+  RCC_APB2PeriphClockCmd(RCC_APB2Periph_ADC1, ENABLE);
+
+  ADC_InitTypeDef adc_settings = {
+    .ADC_Resolution = ADC_Resolution_12b,
+    .ADC_ContinuousConvMode = adc_mode,
+    .ADC_ExternalTrigConvEdge = ADC_ExternalTrigConvEdge_None,
+    .ADC_ExternalTrigConv = ADC_ExternalTrigConv_T1_TRGO,
+    .ADC_DataAlign = ADC_DataAlign_Right,
+    .ADC_ScanDirection = ADC_ScanDirection_Upward,
+  };
+
+  ADC_Init(ADC1, &adc_settings);
+
+  // Calculate the ADC calibration factor
+  ADC_GetCalibrationFactor(ADC1);
+
+  ADC_TempSensorCmd(ENABLE);
+  ADC_VrefintCmd(ENABLE);
+
+  ADC_ContinuousModeCmd(ADC1, adc_mode);
+  while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADCAL)) { }
```

Comment:
According to section 13.4.1 in the datasheet, calibration has to be done while the ADC is still disabled.